### PR TITLE
simulator: support change peer v2

### LINKED ISSUE #5469: simulator: support joint consensus
## Development Task
If we don't support joint consensus, then tikv after v4.0 could not run in the simulator, which led to some feature could not enable, such as hot schedule based query.

### PR DESCRIPTION
**Pull request by HunDunDM**
Signed-off-by: HunDunDM <hundundm@gmail.com>

### What problem does this PR solve?

Issue Number: Close #5469 

### What is changed and how does it work?

Refactor `task.go`

* Split `AddVoter` and `PromoteLearner`
* Merge `AddVoter` and `AddLearner`
  * `PendingPeer` will now be added first, and will be removed after the simulated snap transfer is complete.
* Support original `ChangePeer` in v2 (`ChangePeerV2` and `len(Changes) == 1`)
  * Support `DemoteVoter`, and check if the Peer is not the Leader
* Support enter joint state (`ChangePeerV2` and `len(Changes) > 1`)
* Support leave joint state (`ChangePeerV2` and `len(Changes) == 0`)
* `TransferLeader` adds `Voter` checks
* `TransferLeader` supports the parsing field `Peers` (used to let TiKV find the optimal Peer by itself)
* `RemovePeer` now always also removes the corresponding `DownPeer` and `PendingPeer`

### Check List

Tests

- Manual test (add detailed scripts or steps below)

### Release note

```release-note
None
```

### CODE REVIEW

File: server/core/region_option.go [hunk 1]
```diff
@@ -336,6 +336,39 @@ func WithPromoteLearner(peerID uint64) RegionCreateOption {
 	}
 }
 
+// WithPromoteLearnerEnter promotes the learner to the joint enter stage.
+func WithPromoteLearnerEnter(peerID uint64) RegionCreateOption {
```
Comment (bufferflies) [inline server/core/region_option.go:340]: how about exract one same function `WithRole(peerID unit64,role metapb.PeerRole)`

File: tools/pd-simulator/simulator/task.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2017 TiKV Project Authors.
+// Copyright 2022 TiKV Project Authors.
```
Comment (bufferflies) [inline tools/pd-simulator/simulator/task.go:1]: why change it?

File: tools/pd-simulator/simulator/task.go [hunk 3]
```diff
@@ -49,380 +53,462 @@ const (
...
+	case resp.GetMerge() != nil:
 		targetRegion := resp.GetMerge().GetTarget()
-		return &mergeRegion{
-			regionID:     regionID,
-			epoch:        epoch,
-			targetRegion: targetRegion,
+		desc = fmt.Sprintf("merge region %d into %d", regionID, targetRegion.GetId())
+		op = &mergeRegion{targetRegion: targetRegion}
+	case resp.GetSplitRegion() != nil:
+		// TODO: support split region
```
Comment (lhy1024) [inline tools/pd-simulator/simulator/task.go:116]: when will we support it?

**Review by lhy1024 — COMMENTED** [review]
Do we need some manual test?

File: tools/pd-simulator/simulator/task.go (hunk 2 above)
Comment (HunDunDM) [reply tools/pd-simulator/simulator/task.go:1]: Caused by the newly created file, it has been rolled back.

File: tools/pd-simulator/simulator/task.go (hunk 3 above)
Comment (HunDunDM) [reply tools/pd-simulator/simulator/task.go:116]: We need to discuss how to perform for the three different splits.

**Review by bufferflies — APPROVED** [review]

**Review by nolouch — COMMENTED** [review]
/merge

**Review by nolouch — APPROVED** [review]

**Pull request merged**